### DISCUSSION TITLE
Compute Pipeline. Consecutive storage buffer filling. Is it possible to implement?

### DISCUSSION THREAD

**Discussion by dalduba** [gpuweb/gpuweb#3722] [Q&A] [answered]
Hello. My computational issue has variable result size. This is why I need to fill my storage buffer consecutive from any thread. There is no mutex in WGSL, but there are atomics.
After some research I came up with the following approach. I think it is a kind of spinlock.

```
// ithread_number is calculated from global_id and always greater or equal 0

// try lock until success
while(!atomicCompareExchangeWeak(&thread_id_locker, -1, ithread_number).exchanged) {
    // do nothing
}

// Do here my work: fill the buffer using offset that is also atomic and updates after writing a portion of data

// unlock
atomicStore(&thread_id_locker, -1);
```

If thread_id_locker has storage address space, then it seams that deadlock happen and I have "Device is lost" error. In workgroup address space there are no errors.

I have the following questions:
1. Is it correct way to do what I want? If correct, what I do wrong?
3. If not correct, is there another way to do what I want?
4. If not correct, in which cases should I use atomicCompareExchangeWeak?

**Comment by Kangz** [comment, marked as answer]
I'm not too familiar with the exact memory model of WebGPU, but you should know that memory models on GPU are VERY different than the ones on the CPU, and code that looks like it should work on a CPU might not work on a GPU because cache flushes can be explicit, because execution of some threads is tangled, etc. I'm sure something like that is happening in your case, though it's hard to know what.

The traditional way to allocate space in a storage buffer to store results is just to do an `atomicAdd` that will return to you the start of your allocation and bump the offset for future allocations. That's known to work extremely well and doesn't run into all the issues above because the control flow is much simpler.

**Reply by dalduba** [reply]
Thanks! It's working! It seems my approach was overcomlicated.

But anyway the application for atomicCompareExchangeWeak is not clear.
